File: benchmarks/knucleotide.js
const { performance } = require('perf_hooks');

const t0 = performance.now();
const n = 200000;
const k = 10;
const cs = new Uint8Array(n);
let seed = 42;
const chars = [65, 67, 71, 84]; // A, C, G, T
for (let i = 0; i < n; i++) {
    seed = (seed * 1103515245 + 12345) % 2147483648;
    cs[i] = chars[Math.floor(seed / 65536) % 4];
}
const counts = new Map();
// Simple JS string conversion for map keys is faster than array keys
for (let j = 0; j <= n - k; j++) {
    let key = "";
    for (let l = 0; l < k; l++) {
        key += String.fromCharCode(cs[j + l]);
    }
    counts.set(key, (counts.get(key) || 0) + 1);
}
console.log(counts.size);
console.log(`bench_ns=${Math.round((performance.now() - t0) * 1e6)}`);

File: benchmarks/knucleotide.rs
use std::time::Instant;
use std::collections::HashMap;
use std::str;

fn main() {
    let t0 = Instant::now();
    let n = 200000;
    let k = 10;
    let mut cs = vec![0u8; n];
    let mut seed = 42i64;
    let chars = b"ACGT";
    
    for i in 0..n {
        seed = (seed * 1103515245 + 12345) % 2147483648;
        cs[i] = chars[((seed / 65536) % 4) as usize];
    }
    
    let mut counts: HashMap<&[u8], i32> = HashMap::new();
    for j in 0..=(n - k) {
        *counts.entry(&cs[j..j+k]).or_insert(0) += 1;
    }
    
    println!("{}", counts.len());
    println!("bench_ns={}", t0.elapsed().as_nanos());
}
